<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание" data-toc-modified-id="Описание-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Ответы на вопросы.</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

## Описание

Для проверки платежеспособности нужно использовать таблицу со статистикой о платёжеспособности клиентов банка. Задача: выяснить  влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. 
Сначала предобработаем данные, а затем ответим на поставленные вопросы:

Есть ли зависимость между наличием детей и возвратом кредита в срок?
Есть ли зависимость между семейным положением и возвратом кредита в срок?
Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Как разные цели кредита влияют на его возврат в срок?

## Обзор данных

Для начала импортируем библиотеку Пандас и проверим что данные выгружаются без ошибок:

In [59]:
import pandas as pd
data = pd.read_csv('data.csv')
data.head(15)



children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

In [60]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


использовав метод info() выведем информацию о данных. Видно что в days_employed и total_income есть пропущенные значения. Так же видно что в days_employed есть отрицательные значения, что неправильно. В education есть повторяющиеся значения, как например "ВЫСШЕЕ" и "высшее".
в days_employed пропущенные значения вероятно появились из-за невозможности узнать точный трудовой стаж. Аналогично с total_income, но возможно люди отказались делиться такой информацией.Лучше всего заполнить пропуски в этих столбцах медианным значением потому что это не повлияет на набор общий данных, так как доля пропусков составляет около 10%. 

Узнаем сколько Nan и None в данных:

In [61]:

data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

## Предобработка данных

### Заполнение пропусков

уберем все отрицательные значения days_employed умножив их на единицу:

In [62]:
data.loc[data['days_employed'] < 1 , 'days_employed'] = data.loc[data['days_employed'] < 1 , 'days_employed']*(-1)

рассчитаем медианные значением в столбцах days_employed и total_income:

In [63]:
days_employed_median=data['days_employed'].median()
total_income_median=data['total_income'].median()

print(days_employed_median)
print(total_income_median)
 


2194.220566878695
145017.93753253992


среднее количество трудового стажа составляет примерно 2194 дней, а ежемесечного дохода 145017.

и заменим пропуски медианным значением в столбцах days_employed и total_income:

In [64]:
data['days_employed'] = data['days_employed'].fillna(days_employed_median) 

In [65]:
data['total_income'] = data['total_income'].fillna(total_income_median) 

### Проверка данных на аномалии и исправления.

Проверим что NaN и None больше нет:

In [66]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [67]:
display(data.head(15))

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0    2194.220567         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  145017.937533                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Как видно из таблицы отрицательных значений в days_employed больше нет

### Изменение типов данных.

изменим тип float64 столбца total_income:

In [68]:
data['total_income'] = data['total_income'].astype('int')



И проверим что все изменилось:

In [69]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int32  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 1.9+ MB


### Удаление дубликатов.


применив метод value_counts убедимся что в шкале education есть неявные дубликаты:

In [70]:
print(data['education'].value_counts()) 

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64


заменим все значения к нижнему регистру вызовом метода str.lower:

In [71]:
data['education'] = data['education'].str.lower()


посчитаем общее число дубликатов в таблице:

In [72]:
data.duplicated().sum()

71

удалим дубликаты из данных:

In [73]:
data=data.drop_duplicates()

проверим что дубликаты действительно удалились:

In [74]:
data.duplicated().sum()

0

проверим что неявные дубликаты в столбце education удалены:

In [75]:
print(data['education'].value_counts()) 

среднее                15172
высшее                  5250
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64


аналогично проверим другим столбцы:

In [76]:
print(data['children'].value_counts()) 

 0     14091
 1      4808
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


отрицательное количество детей это аномалия, поэтому заменим значение -1 на 1:

In [77]:
data.loc[data['children'] == -1 , 'children'] = data.loc[data['children'] == -1 , 'children']*(-1)

проверим что -1 больше нет в столбце children:

In [78]:
print(data['children'].value_counts()) 

0     14091
1      4855
2      2052
3       330
20       76
4        41
5         9
Name: children, dtype: int64


Имеется 76 человек у которых 20 детей. Значение является аномальным. Предположу что это ошибка ввода данных и заменю все значения 20 на 2: 
К тому же нет значений больше 5 кроме 20 детей.

In [79]:
data.loc[data["children"] == 20,"children"]= 2


и проверим что значений 20 больше нет:

In [80]:
print(data['children'].value_counts()) 

0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64


In [81]:
print(data['dob_years'].value_counts()) 

35    616
40    607
41    605
34    601
38    597
42    596
33    581
39    572
31    559
36    554
44    545
29    544
30    537
48    536
37    536
50    513
43    512
32    509
49    508
28    503
45    496
27    493
52    484
56    483
47    477
54    476
46    472
53    459
57    456
58    454
51    446
59    443
55    443
26    408
60    374
25    357
61    354
62    348
63    269
24    264
64    260
23    252
65    193
22    183
66    182
67    167
21    111
0     101
68     99
69     85
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


в столбце dob_years обнаружилось 101 человек у которых возраст равен нулю. Возможно часть людей отказалась называть возраст или это ошибка сбора данных. Заменим это значение на медианное. Так как медианное значение равно 42.0 сразу преобразуем его в формат int.

In [82]:
dob_years_median=data['dob_years'].median()
dob_years_median=int(dob_years_median)
data.loc[data['dob_years'] == 0 , 'dob_years'] = dob_years_median

убедимся что нулей больше нет:

In [83]:
print(data['dob_years'].value_counts()) 

42    697
35    616
40    607
41    605
34    601
38    597
33    581
39    572
31    559
36    554
44    545
29    544
30    537
48    536
37    536
50    513
43    512
32    509
49    508
28    503
45    496
27    493
52    484
56    483
47    477
54    476
46    472
53    459
57    456
58    454
51    446
59    443
55    443
26    408
60    374
25    357
61    354
62    348
63    269
24    264
64    260
23    252
65    193
22    183
66    182
67    167
21    111
68     99
69     85
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


проверим что в education_id столько же значений сколько в education

In [84]:
print(data['education_id'].value_counts())

print(data['education'].value_counts()) 

1    15172
0     5250
2      744
3      282
4        6
Name: education_id, dtype: int64
среднее                15172
высшее                  5250
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64


все верно.

аналогично проверим и с family_status и с family_status_id

In [85]:
print(data['family_status'].value_counts())

print(data['family_status_id'].value_counts()) 

женат / замужем          12339
гражданский брак          4151
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64
0    12339
1     4151
4     2810
3     1195
2      959
Name: family_status_id, dtype: int64


In [86]:
print(data['gender'].value_counts())

F      14174
M       7279
XNA        1
Name: gender, dtype: int64


обнаружился один человек с гендером XMA... оставим его на всякий случай :)

In [87]:
print(data['income_type'].value_counts())

сотрудник          11084
компаньон           5078
пенсионер           3829
госслужащий         1457
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64


артефактов в income_type не обнаружено

In [88]:
print(data['debt'].value_counts())

0    19713
1     1741
Name: debt, dtype: int64


0 нет задолженности, 1 есть задолженность. Все верно.

In [89]:
print(data['purpose'].value_counts())

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

для поиска дубликатов я использовал метод value_counts(). Так как он короче и нагляднее чем  duplicated().sum()

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

создадим два новых датафрейма где каждому уникальному значению из education соответствует уникальное значение education_id:

In [90]:
education_row = [
    ['высшее'],
    ['среднее'],
    ['неоконченное высшее'],
    ['начальное'],
    ['ученая степень'],
]
education_columns = ['education']

dataframe_for_education = pd.DataFrame(data=education_row , columns=education_columns)

print(dataframe_for_education)

             education
0               высшее
1              среднее
2  неоконченное высшее
3            начальное
4       ученая степень


и аналогично для family_status и family_status_id:

In [91]:
family_row = [
    ['женат / замужем'],
    ['гражданский брак'],
    ['вдовец / вдова'],
    ['в разводе'],
    ['Не женат / не замужем'],
]
family_columns = ['family']

dataframe_for_family = pd.DataFrame(data=family_row , columns=family_columns) 

print(dataframe_for_family)

                  family
0        женат / замужем
1       гражданский брак
2         вдовец / вдова
3              в разводе
4  Не женат / не замужем


In [92]:
display(data.head(15))

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              среднее             1   
8          2    6929.865299         35               высшее             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0    2194.220567         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        145017   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   

                              purpose  
0                       покупка жилья  
1             приобретение автомобиля  
2                       покупка жилья  
3          дополнительное образование  
4                     сыграть свадьбу  
5                       покупка жилья  
6                   операции с жильем  
7                         образование  
8               на проведение свадьбы  
9             покупка жилья для семьи  
10               покупка недвижимости  
11  покупка коммерческой недвижимости  
12                    сыграть свадьбу  
13            приобретение автомобиля  
14         покупка жилой недвижимости

удалим из исходной таблицы столбцы education и family_status используя метод drop. параметр axis=1 так как мы удаляем столбец:

In [93]:

data=data.drop(columns=['education'],axis = 1)
data=data.drop(columns=['family_status'],axis = 1)




и проверим что столбцы действительно удалились:

In [94]:
display(data.head(15))

children  days_employed  dob_years  education_id  family_status_id gender  \
0          1    8437.673028         42             0                 0      F   
1          1    4024.803754         36             1                 0      F   
2          0    5623.422610         33             1                 0      M   
3          3    4124.747207         32             1                 0      M   
4          0  340266.072047         53             1                 1      F   
5          0     926.185831         27             0                 1      M   
6          0    2879.202052         43             0                 0      F   
7          0     152.779569         50             1                 0      M   
8          2    6929.865299         35             0                 1      F   
9          0    2188.756445         41             1                 0      M   
10         2    4171.483647         36             0                 0      M   
11         0     792.701887         40             1                 0      F   
12         0    2194.220567         65             1                 1      M   
13         0    1846.641941         54             2                 0      F   
14         0    1844.956182         56             0                 1      F   

   income_type  debt  total_income                            purpose  
0    сотрудник     0        253875                      покупка жилья  
1    сотрудник     0        112080            приобретение автомобиля  
2    сотрудник     0        145885                      покупка жилья  
3    сотрудник     0        267628         дополнительное образование  
4    пенсионер     0        158616                    сыграть свадьбу  
5    компаньон     0        255763                      покупка жилья  
6    компаньон     0        240525                  операции с жильем  
7    сотрудник     0        135823                        образование  
8    сотрудник     0         95856              на проведение свадьбы  
9    сотрудник     0        144425            покупка жилья для семьи  
10   компаньон     0        113943               покупка недвижимости  
11   сотрудник     0         77069  покупка коммерческой недвижимости  
12   пенсионер     0        145017                    сыграть свадьбу  
13   сотрудник     0        130458            приобретение автомобиля  
14   компаньон     1        165127         покупка жилой недвижимости

### Категоризация дохода.

In [95]:
def income_category_group(income):  
    if income >= 0 and income <= 30000:
        return 'E'
    if income >= 30001 and income <= 50000:
        return 'D'
    if income >= 50001 and income <= 200000:
        return 'C'
    if income >= 200001 and income <= 1000000:
        return 'B'
    if income >= 1000001 :
        return 'A'
    else:
        return 'unknown'
    
data['total_income_category']=data['total_income'].apply(income_category_group)

проверяем что новый столбец появился в таблице:

In [96]:
display(data.head(10))

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

### Категоризация целей кредита.

In [97]:
def purpose_category_group(purpose):  
    if 'свадьб' in purpose:
        return 'проведение свадьбы'

    if 'недвижимост' in purpose:
        return 'операции с недвижимостью'
    if 'жиль' in purpose:
        return 'операции с недвижимостью'
    
     
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    
    if 'образован' in purpose:
        return 'получение образования'
    
    if 'автом' in purpose:
        return 'операции с автомобилем'
    
    else:
        return 'Другое'
    
    
data['purpose_category']=data['purpose'].apply(purpose_category_group)

проверяем что новый столбец появился в таблице:

In [98]:
display(data.head(10))
#display(data.loc[data['children'] ==20] )


children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы  
5                     B  операции с недвижимостью  
6                     B  операции с недвижимостью  
7                     C     получение образования  
8                     C        проведение свадьбы  
9                     C  операции с недвижимостью

### Ответы на вопросы.

для ответов на вопросы напишем функцию которая возвращает долю должников от общего количества клиентов

In [99]:
def share_debtors(a):
    return (sum(a)/ len(a))

In [100]:
data_pivot_children = data.pivot_table(index='children', values='debt', aggfunc=['count','sum',share_debtors])
print(data_pivot_children)

          count   sum share_debtors
           debt  debt          debt
children                           
0         14091  1063      0.075438
1          4855   445      0.091658
2          2128   202      0.094925
3           330    27      0.081818
4            41     4      0.097561
5             9     0      0.000000


Большая часть людей имеет от 0 до 2 детей и при этом чем больше детей чем вероятность выплаты кредита вовремя уменьшется, т.е. Лучше всего возвращают кредит люди без детей.

Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [101]:
data_pivot_family = data.pivot_table(index='family_status_id', values='debt', aggfunc=['count','sum',share_debtors])
# 0        женат / замужем
# 1       гражданский брак
# 2         вдовец / вдова
# 3              в разводе
# 4  Не женат / не замужем
print(data_pivot_family)


 

                  count  sum share_debtors
                   debt debt          debt
family_status_id                          
0                 12339  931      0.075452
1                  4151  388      0.093471
2                   959   63      0.065693
3                  1195   85      0.071130
4                  2810  274      0.097509


заметим, что "в разводе" и "вдовец / вдова" это тоже "Не женат / не замужем". Поэтому можно объединить эти строчки.

In [102]:
data.loc[data['family_status_id'] == 3 , 'family_status_id'] = 4
data.loc[data['family_status_id'] == 2 , 'family_status_id'] = 4

In [103]:
data_pivot_family = data.pivot_table(index='family_status_id', values='debt', aggfunc=['count','sum',share_debtors])
display(data_pivot_family)

count  sum share_debtors
                   debt debt          debt
family_status_id                          
0                 12339  931      0.075452
1                  4151  388      0.093471
4                  4964  422      0.085012

Выходит что люди состоящие в браке чаще всего вовремя выплачивают кредит, а люди состоящие в гражданском браке реже.

In [104]:
#data_pivot_income_category = data.pivot_table(index=['total_income_category'], values='debt', aggfunc='mean')
data_pivot_income_category = data.pivot_table(index='total_income_category', values='debt', aggfunc=['count','sum',share_debtors])
display(data_pivot_income_category)

count   sum share_debtors
                        debt  debt          debt
total_income_category                           
A                         25     2      0.080000
B                       5041   356      0.070621
C                      16016  1360      0.084915
D                        350    21      0.060000
E                         22     2      0.090909

людей с очень высоким заработком и очень низким очень мало поэтому не будем их учитывать. Самыми надежными надежными являются люди с доходом ниже среднего, но их всего 350. Люди с доходом выше среднего более неджны чем люди со средним доходом.



In [105]:
#data_pivot_purpose_category = data.pivot_table(index=['purpose_category'], values='debt', aggfunc='mean')
data_pivot_purpose_category = data.pivot_table(index='purpose_category', values='debt', aggfunc=['count','sum',share_debtors])
display(data_pivot_purpose_category)

count  sum share_debtors
                           debt debt          debt
purpose_category                                  
операции с автомобилем     4306  403      0.093590
операции с недвижимостью  10811  782      0.072334
получение образования      4013  370      0.092200
проведение свадьбы         2324  186      0.080034

Люди которые берут кредит на недвижимость и на проведение свадьбы выплачивают кредит вовремя чаще чем люди которые берут кредит на операцию с авто и получение образования. На операции с недвижимостью люди берут чаще всего и возвращают чаще всего.

## Общий вывод:

В ходе проекта было проведено исследование надёжности заемщиков клиентов банка. Были пропущены данные в столбцах days_employed и total_income. Пропуски были заменены медианными значениями. Так же были обнаружены ошибка с отрицательным значением количества детей, отрицательным количеством рабочих дней и возрастом равным нулю. Все ошибки были устранены. Были удалены все явные дубликаты и неявные дубликаты в графе education.  
Клиенты были категоризированы по целям кредита и по доходности.

Были проанализированы задолженности клиентов по кредитам в зависимости от количества детей в семье, семейного положения, уровня месячного дохода и цели кредита. Для большей наглядности были построены сводные таблицы.

На вероятность платежей по кредиту в срок влияют такие факторы как цель кредита, семейное положение, количество детей и доход. Банку можно порекомендовать чаще давать кредиты на операции с недвижимостью и проведение свадеб - в таком случае вероятность выплаты кредита больше чем выплата кредита выданного на операции с авто или на получение образования. Так же имеет значение семейное положение: если человек в официально зарегистрированном браке то вероятность выплаты кредита всрок больше чем если человек не женат и еще больше если состоит в гражданском браке. Так же люди с доходом выше среднего (от 200001 до 1000000 в месяц) чаще выплачивают кредит чем люди со средним доходом. Большинство людей имеют от 0 до 2 детей и при этом чем больше детей чем меньше вероятность выплаты кредита в срок, иначе говоря самыми благонадежными клиентами являются клиенты без детей.